File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    res = 0

    def height(self, root) -> int:
        if not root: return 0

        hL = self.height(root.left)
        hR = self.height(root.right)
        self.res = max(self.res, hL+hR)
        
        return 1 + max(hL, hR)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.height(root)
        
        return self.res

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;

    public int height(TreeNode root) {
        if (root == null) return 0;

        int hL = this.height(root.left);
        int hR = this.height(root.right);

        this.res = Math.max(this.res, hL+hR);
        return 1 + Math.max(hL, hR);
    }

    public int diameterOfBinaryTree(TreeNode root) {
        this.height(root);

        return this.res;
    }
}